lib/build.ts

```diff
@@ -187,3 +192,64 @@ export async function runBuildTask(
 		});
 	});
 }
+
+async function doesImageHaveV2Manifest(
+	task: BuildTask,
+	registrySecrets: RegistrySecrets,
+): Promise<boolean> {
+	let tag: string | undefined = task.tag;
+	let imageName: string | undefined = task.imageName;
+
+	if (!imageName) {
+		if (!task.dockerfile) {
+			return false;
+		}
+		const parsedDockerfile = DockerfileParser.parse(task.dockerfile);
+		const dockerInstructions = parsedDockerfile.getInstructions();
+		const fromInstructions = dockerInstructions.filter(
+			(inst) => inst.getKeyword() === 'FROM',
+		);
+		if (fromInstructions.length === 0) {
+			// bail out, no FROM lines found?
+			return false;
+		}
+		const fromLineArguents = fromInstructions[0].getArguments();
+		if (fromLineArguents.length === 0) {
+			// bail out, no image specified?
+			return false;
+		}
+		const fromSource = fromLineArguents[0].getValue();
+		const separatorIndex = fromSource.lastIndexOf(':');
+
+		imageName = fromSource;
+		if (
+			separatorIndex > 0 &&
+			separatorIndex < fromSource.length - 1 &&
+			fromSource[separatorIndex + 1] !== '/'
+		) {
+			// NOTE:  We use only the _first_ FROM directive to determine
+			//        if there is a v2 manifest and therefore we can use
+			//        the `platform` flag.  Mixing base images between
+			//        those that have a v2 manifest and those that do not,
+			//        while specifying `platform` is not supported.
```

Comment:
I remember we discussed that if the Dockerfile had multiple FROM lines and some of them had images with v2 manifests and others did not, we might not be able to get it to work (either the build would fail, or the wrong arch might be selected for some of the images), but perhaps it would still be worth it to detect the mix of v1 and v2 manifest, and return a certain value:
* `true` if we believe that it is preferable to allow the Docker engine to throw an error for old balenalib base images (no v2 manifest). I think this is my preference, assuming that balenalib images are being updated to use v2 manifests, in which case we could tell users to update their balenalib base images to fix the error. In fact, if balenalib images have already been updated (have them?), this could even detect is the image repo is "balenalib" is throw an error message: "Please update the balenalib image to allow the use of multistage Dockerfiles with multiarch images" (only when manifest v1 is mixed with manifest v2).
* `false` if we believe that is preferable to prevent the Docker engine to throw an error but allow wrong architecture to be selected for multiarch images (e.g. Intel instead of ARM).

Reply:
You make some solid points.  I think in the interest of clarity, I am going to use an enumeration for the return value. Let's see what you think of the next iteration.

Reply:
You make some solid points.  Wouldn't we want the same behavior for _all_ v1 cases as well as _mixed_ v1 / v2 cases?  In both, Docker will have to make some guesses.  

All v2 = pass `platform` flag.
Any v1 = If `balenalib` images present, throw error telling them to update.  Else, don't pass `platform` and hope for the best. 

Let's see what you think of the new implementation. 😄

Reply:
FYI, I am on leave the whole of this week, hence not readily replying to your comments and new commits.

Reply:
😆  I wasn't expecting any answers this week, but I thank you anyway. 
> FYI, I am on leave the whole of this week, hence not readily replying to your comments and new commits.